Fullcontact bundle api upgrade

mohit-rocks opened the pull request:
closes https://github.com/mautic/mautic/issues/8213
resolves #8079

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | 
| Automated tests included? | N
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | #8213 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. 
2. 

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. 

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.

mohit-rocks committed:
Added initial support for the integrations bundle.

mohit-rocks committed:
Added initial support for the integrations bundle.

mohit-rocks committed:
Update full contact API calls as per the new API methods.

mohit-rocks committed:
Update full contact API calls as per the new API methods.

mohit-rocks committed:
Update API callback to include key added in integraton configuration.
Modify curl callback to include newly added authorization token.

mohit-rocks committed:
Update API callback to include key added in integraton configuration.
Modify curl callback to include newly added authorization token.

mohit-rocks committed:
Add integration suuport for company and contact bundles.

mohit-rocks committed:
Add integration suuport for company and contact bundles.

mohit-rocks committed:
Adding field mapping support for the data sync.

mohit-rocks committed:
Adding field mapping support for the data sync.

mohit-rocks committed:
Implment contact storage service to store the contacts and perform related changes for the sync.

mohit-rocks committed:
Implment contact storage service to store the contacts and perform related changes for the sync.

mohit-rocks committed:
Implement contact sync and related changes.

mohit-rocks committed:
Implement contact sync and related changes.

mohit-rocks committed:
Added implementation for compnay lookup.

mohit-rocks committed:
Added implementation for compnay lookup.

mohit-rocks committed:
Fix the batch lookup process. Update the logo.
Added validation to check if key is set in value check.
Remove redundant code in field mapping yaml.

mohit-rocks committed:
Fix the batch lookup process. Update the logo.
Added validation to check if key is set in value check.
Remove redundant code in field mapping yaml.

mohit-rocks committed:
Fixing test case failures and update remaining todo items.

mohit-rocks committed:
Fixing test case failures and update remaining todo items.

kuzmany commented:
I did test. 

1.I see this error: The following Mautic fields are required and must be mapped: Email
Cannot match fields in contact tab

![image](https://user-images.githubusercontent.com/462477/83480601-23c22180-a49c-11ea-95ba-5f828e3b3b6c.png)

2. Need support just sync to Mautic direction

![image](https://user-images.githubusercontent.com/462477/83480657-47856780-a49c-11ea-8897-454f07d428e6.png)

mohit-rocks committed:
Add the email field in the sync options.

mohit-rocks committed:
Add the email field in the sync options.

mohit-rocks committed:
Updating code to fix the PHP Rector issues.

mohit-rocks committed:
Updating code to fix the PHP Rector issues.

kuzmany commented on plugins/MauticFullContactBundle/Sync/Mapping/Field/Field.php:
```diff
@@ -0,0 +1,55 @@
+        $this->label      = $field['displayName'];
+        $this->dataType   = $field['type'];
+        $this->isRequired = (bool) ($field['required'] ?? false);
+        $this->isWritable = true;
```
This should be false and then you enable  Sync only to Mautic direction

kuzmany commented on plugins/MauticFullContactBundle/Sync/Mapping/FieldMappings/CompanyFieldMapping.yaml:
```diff
@@ -0,0 +1,61 @@
+    displayName: "Name"
+    name: name
+    type: text
+-
```
Should add required: true to company name

kuzmany commented on plugins/MauticFullContactBundle/Sync/Mapping/FieldMappings/CompanyFieldMapping.yaml:
```diff
@@ -0,0 +1,61 @@
+    displayName: "Email"
+    name: emails
+    type: text
+-
```
Should add required: true

mohit-rocks replied:
This file is just to provide the mapping of source data field and mapping field.
Do we need to parse this attribute and add it on the configuration modal?

kuzmany replied:
You have this attribute in Field.php DTO. 
My suggestion make email/company name required and in the first place.

https://github.com/mautic/mautic/pull/8783/files#diff-b095ba0cad61c1b1c2ad02108d685e1cR22

kuzmany replied:
Still small improvement.
Maybe email and company email as required.

kuzmany reviewed (changes requested):
I comment some parts to improve it
Also could we rework company fetch not by website, but by company name?
https://dashboard.fullcontact.com/api-ref#company-enrichment

mohit-rocks committed:
Update the code that is failing php rector.
Add option to sync only to mautic in the plugin configuration form.
Remove unnecessary code.

mohit-rocks committed:
Update the code that is failing php rector.
Add option to sync only to mautic in the plugin configuration form.
Remove unnecessary code.

RCheesley commented:
@mohit-rocks thanks for all your work on this so far!  It seems we have some Travis issues remaining - can you take a look?  

Please also take a look at the [docs](https://docs.mautic.org/en/plugins/fullcontact) and if anything is missing or can be improved please make a PR.  If you make the PR and mention this issue by including the link in the description they should get connected up!

We will also need this to have automated tests before we can merge - do you need any help with that?

mohit-rocks commented:
@RCheesley 
There are a few suggestions from @kuzmany about implementing lookup based on company name. Considering in Mautic 2 we had lookup based on email only, should we address that in a separate PR after we merge this?

I'll be working on fixing Travis CI issues and I'll try to add a few test cases.

Also, I see that PR has been updated for destination branch as "staging". I believe staging branch is for 2.x release, is it correct?
cc: @npracht

RCheesley commented:
Hi there, staging is now 3.0 so that is correct. 

I think it is important that we look into being able to do a company look up without the website, often you will add the company name but not fully populate the company profile. If we want to add this additional capability that would be in the 3.1 release, as it adds a new feature.

We could make this PR a bug fix (as the plugin is currently broken) with just the existing functionality, and then follow up with the secondary pull request for 3.1 which improves the plugin capability to being able to look up by company name.

I think it would be most useful if we are able to use whatever information we have in order to suggest a match and enrich the contact/company profile with the missing data.

Once the issues are fixed with Travis you should also get a code coverage report which will give us indications as to the test coverage.

mohit-rocks committed:
Remove unused variable declaration.

mohit-rocks committed:
Update the local branch with latest 3.x release.

RCheesley commented:
Given the work needed to complete we are going to push this back to a later release so that there is sufficient time to test thoroughly.

kuzmany commented:
@RCheesley I tested it before, and it works. If you consider add it to 3.0.1 I can review and test it again

RCheesley commented:
Copying response from Slack for info:

> Given that the release is on Monday and this is a big PR without  sufficient testing we decided to move it back to a later release.
> 
> Currently we have the code coverage report only covering the bundles, with 3.0.2 we plan to extend that to also cover plugins so that will also give more information.
> 
> Here is an earlier thread which gives examples of tests: https://mautic.slack.com/archives/CQMKV0RU1/p1592990407457400?thread_ts=1592743794.454500 - the tests need to be included with the PR.

kuzmany committed:
Add required fields to mapping

kuzmany reviewed (dismissed):
Fetch data works for contacts and companies 👍 

![image](https://user-images.githubusercontent.com/462477/88326496-64188000-cd26-11ea-8c67-db7476549149.png)

Small issues:

- Notifications doesn't work. 
- Automatically update on save?  doesn't work because it's commented https://github.com/mautic/mautic/pull/8783/files#diff-ff735a090ee310e6e820f1ae00be1eefR38
- _Sync objects between Mautic and this integration_ and _Select the objects to sync with Mautic_ not apply- still Lookup button is show in contacts/companies
- If I change synch settings and get lookup second time - contact is not update from social cache

mohit-rocks commented:
Thanks for the quick test @kuzmany. I'll review comments over this weekend.

kuzmany commented:
@mohit-rocks 

> Automatically update on save? doesn't work because it's commented #8783 (files)

Maybe we can remove this feature, because it's kind of risky us it on each save entity. 
Also remove it from docs. We can note to docs free account of FullContact contain 1000 API request

RCheesley commented:
Bumping this to 3.1.1 in the hope that we can address the open questions and get the test coverage up by then @mohit-rocks :)

mabumusa1 commented:
@mohit-rocks Any update on this PR, maybe we can merge it this release?

RCheesley commented:
Bumping to 3.1.2 due to automated test coverage needing to be improved.

RCheesley commented:
@cla-bot check

dennisameling committed:
Merge branch 'features' into fullcontact-bundle-api-upgrade

kuzmany committed:
Merge branch '4.x' into fullcontact-bundle-api-upgrade

kuzmany reviewed (dismissed):
We use it in production.
Works 👍

kuzmany commented:
@mohit-rocks can you  resolve CS fixer?

fedys committed:
CS fix

kuzmany commented:
@escopecz shouldn't we close it and remove it in M6? It not work 4 years at least, then  nobody use it.

RCheesley committed:
Merge branch '4.x' into fullcontact-bundle-api-upgrade

jos0405 commented:
I'm closing as it's not relevant in M6/7
